### PR TITLE
Refactor websocket adapters

### PR DESCRIPTION
**Pull request by HerbCaudill**
- Refactors `BrowserWebSocketClientAdapter` and `NodeWSServerAdapter` for readability and maintainability - see comments below. 
- Makes the retry and keepAlive intervals configurable, so we can speed them up in testing and get those processes covered by tests.
- Handles the `leave` message on the server - see comments. 
- Implements the `disconnect` method on the server - see comments.
- Refactors the package's test suites to simplify setup.
- Adds tests to bring this package to 100% coverage.

### CODE REVIEW

**Comment by HerbCaudill** [comment]
Current dependencies on/for this PR:
* main
  * **PR #243** <a href="https://app.graphite.dev/github/pr/automerge/automerge-repo/243?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #244** <a href="https://app.graphite.dev/github/pr/automerge/automerge-repo/244?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #242** <a href="https://app.graphite.dev/github/pr/automerge/automerge-repo/242?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #248** <a href="https://app.graphite.dev/github/pr/automerge/automerge-repo/248?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/automerge/automerge-repo/248?utm_source=stack-comment).

**Comment by HerbCaudill** [comment]
merged with other stuff

File: packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import {
   NetworkAdapter,
-  type PeerId,
```
Comment (pvh) [inline packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts:3]: These type assertions allow imported type removal for transpilers without loading the full object graph. Is there a reason you took them out?

File: packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts [hunk 2]
```diff
@@ -12,173 +13,162 @@ import {
...
-      throw new Error("WTF, get a socket")
+  onError = (event: WebSocket.ErrorEvent) => {
+    const { code } = event.error
+    if (code === "ECONNREFUSED") {
+      this.#log("Connection refused, retrying...")
+    } else {
+      /* c8 ignore next */
+      throw event.error
     }
+  }
```
[the remark this answers is not in the document]
Comment (pvh) [reply packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts:112]: Much nicer, thanks. It's possible we'll want an exponential backoff at some point but I don't think it's critical at the moment.

File: packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts [hunk 3]
```diff
@@ -12,173 +13,162 @@ import {
...
   }
 
   disconnect() {
-    if (!this.socket) {
-      throw new Error("WTF, get a socket")
-    }
-    this.send({ type: "leave", senderId: this.peerId! })
+    assert(this.peerId)
+    assert(this.socket)
+    this.send({ type: "leave", senderId: this.peerId })
```
[the remark this answers is not in the document]
Comment (pvh) [reply packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts:133]: Will these messages crash an old server?

[the remark this answers is not in the document]
Comment (pvh) [reply packages/automerge-repo-network-websocket/src/BrowserWebSocketClientAdapter.ts:133]: (We should test unrecognized message handling.)

**Pull request closed without merging**